keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,884 @@
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+There are [many
+reasons](http://bit.ly/k8s-multicluster-conversation-starter-doc) why a K8s user
+may want to split their deployments across multiple clusters, but still retain
+mutual dependencies between workloads running in those clusters. Today the
+cluster is a hard boundary, and a service is opaque to a remote K8s consumer
+that would otherwise be able to make use of metadata (e.g. endpoint toplogy) to
+better direct traffic. To support failover or temporarily during migration,
+users may want to consume services spread across clusters, but today that
+requires non-trivial bespoke solutions.
+
+The Multi-Cluster Services API aims to fix these problems.
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+- Define a minimal API to support service discovery and consumption across clusters.
+  - Consume a service in another cluster.
+  - Consume a service deployed in multiple clusters as a single service.
+- When a service is consumed from another cluster its behavior should be
+  predictable and consistent with how it would be consumed within its own
+  cluster.
+- Create building blocks for multi-cluster tooling.
+- Support multiple implementations.
+- Leave room for future extension and new use cases.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Define specific implementation details beyond general API behavior.
+- Change behavior of single cluster services in any way.
+- Define what NetworkPolicy means for multi-cluster services.
+- Solve mechanics of multi-cluster service orchestration.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+#### Terminology
+
+- **supercluster** - A placeholder name for a group of clusters with a high
+  degree of mutual trust and shared ownership that share services amongst
+  themselves. Membership in a supercluster is symmetric and transitive. The set
+  of member clusters are mutually aware, and agree about their collective
+  association.
+- **mcsd-controller** - A controller that syncs services across clusters and
+  makes them available for multi-cluster service discovery (MCSD) and
+  connectivitiy. There may be multiple implementations, this doc describes
+  expected common behavior.
+
+We propose a new CRD called `ServiceExport`, used to specify which services
+should be exposed across all clusters in the supercluster. `ServiceExports` must
+be created in each cluster that the underlying `Service` resides in. Creation of
```

Comment:
The `Service` object is a key piece of this KEP but I didn't find any mention to the `Type` of `Service`. Is this KEP addressing any Type of service?
* ClusterIP: Headless and No-Headless
* NodePort
* LoadBalancer
* ExternalName

My impression after reading it and for the mentions to the EndpoinSlices and the `SuperClusterIP` is that is targeting `ClusterIP` services only, is that correct? 😅

Reply:
You're right, this needs some more clarification. This is how I see it:

The resulting `SuperclusterIP` service that's imported into each cluster is like `ClusterIP` (no node port, no external LB IP), but the source services don't have to be. The supercluster endpoints are the union of endpoints attached to services exported from each cluster, regardless of how those endpoints are used in-cluster.

Let's start with exporting `ClusterIP` - this is the the straightforward case most of the doc assumes. Each `EndpointSlice` associated with the exported service is combined with slices from other clusters to make up the supercluster service. They will be imported to the cluster behind the supercluster IP.

Headless - Today's headless services likely don't want a VIP and may not function properly behind one - there's a note a little further down about stateful sets and complications that come with them. We haven't answered what that could look like. It probably doesn't make sense to export a current headless service to the supercluster, it would work, but likely not the way you want.

`NodePort` and `LoadBalancer` type services both create `ClusterIP` services and that's what would sync. For example If you export a `NodePort` service, the resulting cross-cluster service will still be a `SuperclusterIP` type. You could use node ports to access the cluster-local service in the source cluster, but not in any other cluster, and it would only route to local endpoints.

`ExternalName` really is the outlier here - it doesn't make sense to export an `ExternalName` service. They can't be merged with other exports, and it seems like it would only complicate deployments by even attempting to stretch them across clusters. Why not just create regular `ExternalName` services in each cluster that needs them? Perhaps we should just disallow exporting them. What do you think?

Reply:
That would make sense, but mechanically what does that mean? If "headless" is derived from cluster-level services, is a supercluster service headless if any component service is headless? Only if all are? I was thinking along these lines, e.g. we create pod A/AAAA records based on exporting service headlessness, but at which point should we switch the service record from VIP to pods IPs? I don't really see a case for mixing and matching headless/non-headless services so maybe we can just say a supercluster service is headless if all component services are headless

Reply:
I think we can define resolution, along the lines of "If any same-name exported cluster service has a VIP, the supercluster service also has a VIP.  If all same-name exported cluster services are headless, the supercluster service is headless".

You need to be prepared to allocate a supercluster IP if one member service changes, though.

Reply:
and every non-correct ServiceExport needs to get status, and maybe events.

Reply:
Yeah, that makes sense to me. Adding to the KEP

Reply:
Is there a built-in assumption that all clusters participating in the **supercluster** have the ability to resolve clusterIPs services (not exposed as NodePort or LB services) and/or clusterIPs of pods between each other? Can member clusters continue to have separate network overlays but allow private NodePort/LB service routing?

Reply:
The current assumption is that all participating clusters have mutual connectivity at the Pod level - clusters do not need to resolve each other's clusterIPs, but do need to access each others Pods. The SuperclusterIP introduced here is a VIP, local to the importing cluster, that routes to Pods from exporting clusters.

Connecting directly to a Service LB would start to break down if multiple clusters exported the same service. You'd have multiple LB IPs, one for each cluster, but since there's no concept of weight in kube-proxy, each IP would get an even share of traffic regardless of the number of backing pods. However, I could see configurations where only the specific pods behind an exported service are exposed through a gateway - the work to update [Service APIs](https://github.com/kubernetes-sigs/service-apis) should help with that